### PR TITLE
crimson/thread: generalize Task so it works w/ func returns void

### PR DESCRIPTION
**Pull request by tchaikov**
Signed-off-by: Kefu Chai <kchai@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
@liu-chunmei

**Comment by tchaikov** [comment]
```
05:30:02 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 11.3768/12.1455  => rejected
05:30:02 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 2912.0/3109.0  => rejected
05:30:02 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.50)):: 62.3859/205.249  => accepted
05:30:02 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00545786/0.00512814  => rejected
05:30:02 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 11.3982/12.1008  => rejected
05:30:02 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 2917.0/3097.0  => rejected
05:30:02 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.50)):: 91.4349/216.264  => accepted
05:30:02 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00548083/0.00515751  => rejected
05:30:02 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 49.7752/52.6454  => rejected
05:30:02 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 12742.0/13477.0  => rejected
05:30:02 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.50)):: 1230.96/178.111  => rejected
05:30:02 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.0012481/0.0011791  => rejected
05:30:02 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 49.5565/52.557  => rejected
05:30:02 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 12686.0/13454.0  => rejected
05:30:02 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.50)):: 1249.27/165.393  => rejected
05:30:02 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00125271/0.00118187  => rejected
05:30:02 - WARNING  - cbt      - 14 tests failed out of 16
```

**Comment by tchaikov** [comment]
jenkins test crimson perf

File: src/crimson/thread/ThreadPool.h [hunk 1]
```diff
@@ -22,26 +22,37 @@ struct WorkItem {
...
 
-template<typename Func, typename T = std::invoke_result_t<Func>>
+template<typename Func>
 struct Task final : WorkItem {
   Func func;
-  seastar::future_state<T> state;
+  using T = std::invoke_result_t<Func>;
+  using futurator_t = seastar::futurize<T>;
+  typename futurator_t::value_type value;
+  std::exception_ptr e;
```
Comment (rzarzynski) [inline src/crimson/thread/ThreadPool.h]: The downside might be a slightly bigger memory footprint for holding `void` as `future_state` uses EBO to knock out `std::tuple<>`. Do we have any performance constraints on the `Task`?

File: src/test/crimson/test_thread_pool.cc [hunk 2]
```diff
@@ -25,13 +25,21 @@ seastar::future<> test_accumulate(ThreadPool& tp) {
...
   ThreadPool tp{2, 128, 0};
   seastar::app_template app;
   return app.run(argc, argv, [&tp] {
       return tp.start().then([&tp] {
           return test_accumulate(tp);
+        }).then([&tp] {
+          return test_void_return(tp);
         }).handle_exception([](auto e) {
           std::cerr << "Error: " << e << std::endl;
           seastar::engine().exit(1);
```
Comment (rzarzynski) [inline src/test/crimson/test_thread_pool.cc]: There is a typo in the commit's title (*generize*).

File: src/crimson/thread/ThreadPool.h (hunk 1 above)
Comment (tchaikov) [reply src/crimson/thread/ThreadPool.h]: fixed and repushed.

File: src/test/crimson/test_thread_pool.cc (hunk 2 above)
Comment (tchaikov) [reply src/test/crimson/test_thread_pool.cc]: fixed.

**Comment by tchaikov** [comment]
changelog

- keep using `future_state`
- fixed the typo in the title

**Pull request merged**